File: app/src/main/java/com/example/macarus0/walkiewalkie/view/SelectDogActivity.java
package com.example.macarus0.walkiewalkie.view;

import android.app.Activity;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;

import com.example.macarus0.walkiewalkie.R;
import com.example.macarus0.walkiewalkie.viewmodel.WalkieViewModel;

import butterknife.BindView;
import butterknife.ButterKnife;

public class SelectDogActivity extends AppCompatActivity implements DogListAdapter.DogClickHandler{

    public static final String DOG_ID = "dog_id";

    WalkieViewModel mViewModel;
    @BindView(R.id.select_contact_toolbar)
    Toolbar mToolbar;
    @BindView(R.id.select_items_list)
    RecyclerView mItemsRecyclerView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_select_contact);
        ButterKnife.bind(this);
        mViewModel = ViewModelProviders.of(this).get(WalkieViewModel.class);
        setSupportActionBar(mToolbar);
        // add back arrow to toolbar
        if (getSupportActionBar() != null){
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setDisplayShowHomeEnabled(true);
            getSupportActionBar().setTitle(R.string.select_dog);
        }

        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
        mItemsRecyclerView.setLayoutManager(linearLayoutManager);

        DogListAdapter dogListAdapter = new DogListAdapter();
        dogListAdapter.setDogClickHandler(this::dogClick);
        dogListAdapter.setShowOwners(false);
        mItemsRecyclerView.setAdapter(dogListAdapter);
        mViewModel.getAllAvailableDogs().observe(this, dogListAdapter::setDogs);
    }

    @Override
    public void dogClick(long id) {
        Intent intent = new Intent();
        intent.putExtra(DOG_ID, id);
        setResult(Activity.RESULT_OK, intent);
        finishAfterTransition();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // handle arrow click here
        if (item.getItemId() == android.R.id.home) {
            finishAfterTransition();
        }

        return super.onOptionsItemSelected(item);
    }

}

File: app/src/main/java/com/example/macarus0/walkiewalkie/view/SelectDogActivity.kt
package com.example.macarus0.walkiewalkie.view

import android.app.Activity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.appcompat.widget.Toolbar
import android.view.MenuItem

import com.example.macarus0.walkiewalkie.R
import com.example.macarus0.walkiewalkie.viewmodel.WalkieViewModel

import butterknife.BindView
import butterknife.ButterKnife
import com.example.macarus0.walkiewalkie.data.Dog

class SelectDogActivity : AppCompatActivity() {

    private lateinit var mViewModel: WalkieViewModel
    @BindView(R.id.select_contact_toolbar)
    @JvmField  var mToolbar: Toolbar? = null
    @BindView(R.id.select_items_list)
    @JvmField  var mItemsRecyclerView: RecyclerView? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_select_contact)
        ButterKnife.bind(this)
        mViewModel = ViewModelProviders.of(this).get(WalkieViewModel::class.java)
        setSupportActionBar(mToolbar)
        // add back arrow to toolbar
        if (supportActionBar != null) {
            supportActionBar!!.setDisplayHomeAsUpEnabled(true)
            supportActionBar!!.setDisplayShowHomeEnabled(true)
            supportActionBar!!.setTitle(R.string.select_dog)
        }

        val intent = intent
        val ownerId = intent.getLongExtra(OWNER_ID, -1)

        val linearLayoutManager = LinearLayoutManager(this)
        mItemsRecyclerView!!.layoutManager = linearLayoutManager

        val dogListAdapter = DogListAdapter()
        dogListAdapter.dogClickHandler = {it ->  this.dogClick(it) }
        dogListAdapter.setShowOwners(false)
        mItemsRecyclerView!!.adapter = dogListAdapter
        mViewModel.getAllAvailableDogs(ownerId).observe(this, Observer<List<Dog>> { dogListAdapter.setDogs(requireNotNull(it)) })
    }

    fun dogClick(id: Long) {
        val intent = Intent()
        intent.putExtra(DOG_ID, id)
        setResult(Activity.RESULT_OK, intent)
        finishAfterTransition()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // handle arrow click here
        if (item.itemId == android.R.id.home) {
            finishAfterTransition()
        }

        return super.onOptionsItemSelected(item)
    }

    companion object {

        const val DOG_ID = "dog_id"
        const val OWNER_ID = "owner_id"
    }

}
